wandb/sdk/wandb_settings.py

```diff
@@ -687,6 +699,10 @@ def _update(self, __d=None, _source=None, _override=None, **kwargs):
                 self._check_invalid(k, check[k])
         for data in d, kwargs:
             for k, v in six.iteritems(data):
+                if k not in self.__dict__:
+                    raise KeyError(k)
+                v = self._perform_preprocess(k, v)
+                self._check_invalid(k, v)
```

Comment:
Hmm, does this mean we're calling `_check_invalid` twice for every key?  Could be pre-mature optimization but another option would be to call _check_invalid in _perform_preprocess only if we preprocess the key...

Reply:
agreed, pre-process should go before the first check (first loop).  i guess this would involve either updating the value in place or making a dict copy to be applied in the later in the second loop.

Or maybe there is a better way to structure it, but the idea was that thee update would fail if any k,v failed a check.

Reply:
Good call. I moved the check to the first loop and construct a new dict along the way as a copy. this new dict is used in the subsequent loop.